### ISSUE TITLE
[core] Story Server Filter Via Props is broken

### ISSUE THREAD

**Issue by flaviendelangle** [mui-org/material-ui-x#2013]
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

The story currently crashes because `onFilterModelChange` is called with 0 items and the callback `onFilterChange` assumes that their is one item.

![image](https://user-images.githubusercontent.com/3309670/124261163-c3833080-db30-11eb-9814-8d9d3d746255.png)

## Expected Behavior 🤔

The Story should not crash

## Steps to Reproduce 🕹

Go to [This Story](https://material-ui-x.netlify.app/storybook/?path=/story/x-grid-tests-filter--server-filter-via-props&globals=measureEnabled:false)

**Comment by flaviendelangle** [comment]
@dtassone will this be fixed with your rework of the controlled pattern ? (#1823 )

**Comment by oliviertassinari** [comment]
I'm replacing the "docs" label as we use the storybook for internal purposes. It's not meant to be developers facing.

**Comment by dtassone** [comment]
@flaviendelangle no it won't. The fix for the story looks straightforward, you just need to add null/length checks on `filterModel.items[0].value`

**Comment by flaviendelangle** [comment]
If I just checks if there is an item in `filterModel.items`, then I won't be able to differentiate 

1) an invalid callback with 0 item (= the grid calls `onFilterModelChange` with 0 item even if on the UI their is a filter selected)
2) a valid callback with 0 item (= the grid calls `onFilterModelChange` with 0 item and on the UI their is no filter selected)

Because right now, the story render `XGrid` for the 1st time with a controlled `filterModel` with 1 item.
But `onFilterModelChange` is called right away with 0 items.

```
Filter model in Story render {
  "items": [
    {
      "id": 123,
      "columnField": "commodity",
      "value": "soy",
      "operatorValue": "contains"
    }
  ]
}
Filter model in Story onFilterModelChange {
  "items": [],
  "linkOperator": "and"
}
```

For me the issue is that [applyFilterLinkOperator](https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/hooks/features/filter/useGridFilter.ts#L256) publishes the `GRID_FILTER_MODEL_CHANGE` even when called from [setFilterModel](https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/hooks/features/filter/useGridFilter.ts#L271).
If I add a `emit?: boolean = true` param to `applyFilterLinkOperator`, the story works correctly.

But it seems quite related to the controlled behavior changes you are making so I prefered to ask before :+1:

**Comment by dtassone** [comment]
Ok thanks for investigating, then let's fix it as part of #1909

**Comment by flaviendelangle** [comment]
https://github.com/mui-org/material-ui-x/pull/2080/files

I pushed my code to show the issue
But I can close it :+1:
